src/mpi/init/init_util.c

```diff
@@ -21,12 +21,15 @@ typedef struct Finalize_func_t {
 /* When full debugging is enabled, each MPI handle type has a finalize handler
    installed to detect unfreed handles.  */
 #define MAX_FINALIZE_FUNC 64
+static MPL_initlock_t fstack_lock = MPL_INITLOCK_INITIALIZER;
 static Finalize_func_t fstack[MAX_FINALIZE_FUNC];
 static int fstack_sp = 0;
 static int fstack_max_priority = 0;
 
 void MPIR_Add_finalize(int (*f) (void *), void *extra_data, int priority)
 {
+    /* MPIR_Add_finalize may be called both inside and outside init */
+    MPL_initlock_lock(&fstack_lock);
```

Comment:
This should behave as a global lock, right? Could it come in the critical path and impact the intended concurrency for the per-vci mode?

Reply:
Okay. Then this PR looks good to me.